### DISCUSSION TITLE
Include only code blocks from a .qmd file

### DISCUSSION THREAD

**Discussion by binary10** [quarto-dev/quarto-cli#6378] [Feature Requests]
### Description

Please refer to this documentation:
https://quarto.org/docs/reference/cells/cells-knitr.html#include

Currently, it's possible to pass the following parameter to a code block to include and evaluate code cells in a `.qmd` file. 

```{R}
#| child: script.qmd
```

If this file has only code chunks, then this runs the code appropriately. However, if there is markdown in the file, then the markdown is also included. 

A feature that appears to be missing is to specify that **only** code chunks from a `.qmd` will be included and evaluated. In my workflow, there are times when I would only like to include the code chunks, for example to make some data or variables available to the document at the beginning, while still providing additional narrative around the code blocks in the original file so that the code is well documented and will work as a stand alone page.

Aside from this, I've also tried to use the following feature to accomplish this behavior:
https://quarto.org/docs/authoring/includes.html#computations

This does evaluate the code blocks, however, again, the markdown between the code is included.

**Comment by baptiste** [comment]
One workflow that you might want to try with knitr is [code externalisation](https://yihui.org/knitr/demo/externalization/): your chunks (all of them or just a subset) live in a R file, and the qmd files import them by label.

**Reply by baptiste** [reply]
Technically, I guess one could call `purl()` on the child `script.qmd` to produce a R file with only the labelled code chunks, and use `read_chunks()` on that temporary file to achieve what you're describing (ignore any markdown, only import code chunks).

**Reply by cderv** [reply]
Yes this is exactly how do this currently. 
Doc for `read_chunk()` is here: https://bookdown.org/yihui/rmarkdown-cookbook/read-chunk.html

There is a plan to support reading directly from Rmd / Qmd the chunk labels

* https://github.com/yihui/knitr/issues/2041

**Comment by mcanouil** [comment]
Note that the described behaviour is not Quarto's behaviour but `knitr`'s.
Also `child` option is not the recommended include approach in Quarto as the markdown in it won't be fully processed by Quarto because processed by `knitr`.

As mentioned, `knitr` offers various ways to read the code from a file which let you do whatever you want however you want to do it (see <https://bookdown.org/yihui/rmarkdown/>).

Maybe the `embed` shortcode used for Jupyter Notebooks could be extended to work the same way on Quarto documents.
(_The `include` shortcode (or `knitr` `child` option) purposes is not to read and evaluate code only, but to include everything._)

**Reply by cderv** [reply]
> Maybe the embed shortcode used for Jupyter Notebooks could be extended to work the same way on Quarto documents.
(The include shortcode (or knitr child option) purposes is not to read and evaluate code only, but to include everything.)

FWIW there is a plan to support reading directly from Rmd / Qmd the chunk labels

* https://github.com/yihui/knitr/issues/2041

Otherwise as described above this is currently possible with `purl()` + `read_chunk()`

**Comment by cderv** [comment]
> A feature that appears to be missing is to specify that only code chunks from a .qmd will be included and evaluated. In my workflow, there are times when I would only like to include the code chunks, for example to make some data or variables available to the document at the beginning, while still providing additional narrative around the code blocks in the original file so that the code is well documented and will work as a stand alone page.

Some additional comment in addition to all that was given above. In addition to using [`purl()`](https://bookdown.org/yihui/rmarkdown-cookbook/purl.html) + [`read_chunk()`](https://bookdown.org/yihui/rmarkdown-cookbook/read-chunk.html), you could also use a .R script to store your chunk content and use the `file` chunk option in **knitr** to insert this R script to file the code chunk content. You could reuse the same R script in both document using this simple option. https://bookdown.org/yihui/rmarkdown-cookbook/option-code.html

**Reply by binary10** [reply]
Thanks for the thoughts and the workaround for my particular case. 

It seems to me that supporting the following syntax would be beneficial to run only code from a Quarto file:
````
```{R}
#| file: script.qmd
```
````

This syntax wouldn't need to support including markdown since `{{< include script.qmd >}}` supports loading the entire file, code and markdown. 

The downside of the code block above is that a Quarto file could have code blocks written in multiple environments, however specifying `{R}` could filter the code blocks in the Quarto file for only the `R` code blocks.
